### PR TITLE
Document deprecation warnings feature

### PR DESCRIPTION
**Pull request by liggitt**
Doc for 1.19 portion of:
* https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings
* https://github.com/kubernetes/enhancements/issues/1693
* https://github.com/kubernetes/kubernetes/pull/73032
* kubernetes/kubernetes#92329
* kubernetes/kubernetes#92667

Preview links:
* https://deploy-preview-21581--kubernetes-io-vnext-staging.netlify.app/docs/reference/using-api/deprecation-policy/#rest-resources-aka-api-objects
* https://deploy-preview-21581--kubernetes-io-vnext-staging.netlify.app/docs/reference/access-authn-authz/extensible-admission-controllers/#admissionreview-response-warning
* https://deploy-preview-21581--kubernetes-io-vnext-staging.netlify.app/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation

### CODE REVIEW

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 1]
```diff
@@ -298,6 +298,15 @@ exist and function in releases up to and including X+8.  Only in release X+9,
 when API v1 has aged out, does the Widget resource cease to exist, and the
 behavior get removed.
 
+Starting in Kubernetes 1.19, use of a deprecated REST API:
+
+1. returns a `Warning` header (as defined in RFC7234, Section 5.5) in the API response.
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-policy.md:303]: Optionally:
```suggestion
1. returns a `Warning` header (as defined in [RFC7234](https://tools.ietf.org/html/rfc7234), Section 5.5) in the API response.
```

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 2]
```diff
@@ -298,6 +298,15 @@ exist and function in releases up to and including X+8.  Only in release X+9,
...
+2. sets an `apiserver_requested_deprecated_apis` gauge metric to `1` in the `kube-apiserver` 
+   process with labels for `group`, `version`, `resource`, `subresource` that can be joined
+   to the `apiserver_request_total` metric, and a `removed_version` label that indicates the 
+   Kubernetes release in which the API will no longer be served.
+3. adds a `kubernetes.io/deprecated` audit annotation to the audit event recorded for the request
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-policy.md:308]:
```suggestion
3. adds a `kubernetes.io/deprecated` audit annotation to the audit event recorded for the request.
```

File: content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning.md [hunk 3]
```diff
@@ -274,6 +274,87 @@ version sort order is `v1`, followed by `v1beta1`. This causes the kubectl
 command to use `v1` as the default version unless the provided object specifies
 the version.
 
+### Version deprecation
+
+Starting in v1.19, a CustomResourceDefinition can indicate a version is deprecated.
+When API requests to a deprecated version are made, a warning message is returned in the API response as a header.
+The warning message for each deprecated version can be customized if desired.
```
Comment (deads2k) [inline content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning.md:281]: These are deprecated versions for a particular CR, right?  For instance, I could have foo.v1beta1.group and bar.v1beta1.group and I can choose to only deprecate foo.v1beta.group not the entire group/v1beta1.

**Comment by deads2k** [comment]
one question about CRDs.

File: content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning.md (hunk 3 above)
Comment (liggitt) [reply content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning.md:281]: correct

Comment (liggitt) [reply content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning.md]: updated/clarified

File: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md [hunk 4]
```diff
@@ -628,6 +628,47 @@ So a webhook response to add that label would be:
 {{% /tab %}}
 {{< /tabs >}}
 
+Starting in v1.19, admission webhooks can optionally return warning messages that are returned to the requesting client.
+Warnings can be sent with allowed or rejected admission responses.
+No "Warning:" prefix should be included in the message.
+Warning messages should describe a problem the client making the API request should correct or be aware of.
```
Comment (kbhawkey) [inline content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md:634]: nits: Some possible rewording, capitalization(?):
Starting in Kubernetes v1.19, Admission Webhooks optionally return warning messages to client requests [ in responses to clients].

nit: a bit confusing: `No "Warning:" prefix should be included in the message.`

nit: another possible reword:
Warning messages describe a problem which the client should correct or be aware of.

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 5]
```diff
@@ -298,6 +298,18 @@ exist and function in releases up to and including X+8.  Only in release X+9,
...
 behavior get removed.
 
+Starting in Kubernetes 1.19, making an API request to a deprecated REST API endpoint:
+
+1. returns a `Warning` header (as defined in [RFC7234, Section 5.5](https://tools.ietf.org/html/rfc7234#section-5.5)) in the API response.
+2. adds a `"k8s.io/deprecated":"true"` annotation to the [audit event](/docs/tasks/debug-application-cluster/audit/) recorded for the request.
+3. sets an `apiserver_requested_deprecated_apis` gauge metric to `1` in the `kube-apiserver` 
+   process. The metric has labels for `group`, `version`, `resource`, `subresource` that can be joined
+   to the `apiserver_request_total` metric, and a `removed_version` label that indicates the 
+   Kubernetes release in which the API will no longer be served. The following prometheus query
```
Comment (kbhawkey) [inline content/en/docs/reference/using-api/deprecation-policy.md:308]: nit (v1.19 vs. 1.19): Starting with Kubernetes v1.19 ...
nit: The following Prometheus query returns information ...

**Comment by sftim** [comment]
/retitle Add deprecation warnings relevant to v1.19
(hope that retitling is OK)

**Comment by sftim** [comment]
Sorry for the spam.
/retitle Document deprecation warnings feature

**Comment by annajung** [comment]
Due to @liggitt being out, the release team decided to create a new PR to make changes required from the feedback above to get this doc merged in for 1.19 release. 

https://github.com/kubernetes/website/pull/22517

File: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md [hunk 6]
```diff
@@ -628,6 +628,47 @@ So a webhook response to add that label would be:
...
+Limit warnings to 120 characters if possible. Warnings over 256 characters and large numbers of warnings may be truncated.
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md:635]: It took me a long time to work out the intended meaning. I think this is equivalent.

```suggestion
If you're implementing a webhook that returns a warning:

- don't include the header name `Warning:` in the message
- write messages that describe the problem the client making the API request should correct or be aware of
- limit messages to 120 characters if possible.

{{< caution >}}
The API server might truncate a warning message that exceeds 256 characters.  
Also, if one or more webhooks issue several warnings that combine to exceed 256 characters, the API server might truncate a shorter message that takes the combined warning-text length over the limit.
{{< /caution >}}
```

File: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md [hunk 7]
```diff
@@ -628,6 +628,47 @@ So a webhook response to add that label would be:
 {{% /tab %}}
 {{< /tabs >}}
 
+Starting in v1.19, admission webhooks can optionally return warning messages that are returned to the requesting client.
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md:631]:
```suggestion
Starting in v1.19, admission webhooks can optionally return warning messages that are returned to the requesting client in HTTP Warning: headers, with a warning code of 299.
```
(is that correct? I haven't checked the actual code).

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 8]
```diff
@@ -298,6 +298,18 @@ exist and function in releases up to and including X+8.  Only in release X+9,
...
+   would return information about requests made to deprecated APIs which will be removed in v1.22:
+   
+   `apiserver_requested_deprecated_apis{removed_version="1.22"} * on(group,version,resource,subresource) group_right() apiserver_request_total`
```
Comment (sftim) [inline content/en/docs/reference/using-api/deprecation-policy.md:311]: Nit: I would also use a code block marked as promql (even though I'm not sure that the website currently knows how to syntax highlight PromQL).

File: content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning.md [hunk 9]
```diff
@@ -274,6 +274,87 @@ version sort order is `v1`, followed by `v1beta1`. This causes the kubectl
 command to use `v1` as the default version unless the provided object specifies
 the version.
 
+### Version deprecation
+
+Starting in v1.19, a CustomResourceDefinition can indicate a particular version of the resource it defines is deprecated.
```
Comment (sftim) [inline content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning.md:279]:
```suggestion

{{< feature-state state="stable" for_k8s_version="v1.19" >}}
 
Starting in v1.19, a CustomResourceDefinition can indicate that a particular version of the resource it defines is deprecated.
```

**Comment by annajung** [comment]
@liggitt reached out and mentioned that he will be available to review the comments and move the doc forward for the release before the deadline. Closing the replacement PR https://github.com/kubernetes/website/pull/22517 that was created. Sorry for the confusion everyone! The release team took the action necessary to make sure this gets merged in on time.

File: content/en/docs/reference/using-api/deprecation-policy.md (hunk 5 above)
Comment (liggitt) [reply content/en/docs/reference/using-api/deprecation-policy.md:308]: > nit (v1.19 vs. 1.19): Starting with Kubernetes v1.19 ...

Happy to follow whatever convention the docs team wants, but there are 5 existing instances of "Starting in v1..." and 1 existing instance of "Starting in 1..." in the docs

> nit: The following Prometheus query returns information ...

done

File: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md (hunk 6 above)
Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md:635]: that's not quite equivalent, but I tried to adapt to match the suggested structure

File: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md (hunk 7 above)
Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md:631]: correct

File: content/en/docs/reference/using-api/deprecation-policy.md (hunk 8 above)
Comment (liggitt) [reply content/en/docs/reference/using-api/deprecation-policy.md:311]: done

**Comment by liggitt** [comment]
rebased on dev-1.19 and pushed a commit addressing the doc review comments

File: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md (hunk 4 above)
Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md:634]: updated in the restructured version suggested in https://github.com/kubernetes/website/pull/21581#discussion_r454635053

File: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md (hunk 6 above)
Comment (kbhawkey) [reply content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md:635]: Can you correct:
`write messages that describe the problem the client making the API request should correct or be aware of`
This is hard to parse.

**Comment by savitharaghunathan** [comment]
@liggitt does it need a tech lgtm as well?

**Comment by sftim** [comment]
Not sure if you already committed and (force-)pushed @liggitt but GitHub isn't showing me any recent changes.

**Comment by liggitt** [comment]
I did push... looks like github is having issues...

@deads2k reviewed in https://github.com/kubernetes/website/pull/21581#issuecomment-654388559

File: content/en/docs/reference/using-api/deprecation-policy.md [hunk 10]
```diff
@@ -298,6 +298,18 @@ exist and function in releases up to and including X+8.  Only in release X+9,
 when API v1 has aged out, does the Widget resource cease to exist, and the
 behavior get removed.
 
+Starting in Kubernetes 1.19, making an API request to a deprecated REST API endpoint:
+
+1. returns a `Warning` header (as defined in [RFC7234, Section 5.5](https://tools.ietf.org/html/rfc7234#section-5.5)) in the API response.
+2. adds a `"k8s.io/deprecated":"true"` annotation to the [audit event](/docs/tasks/debug-application-cluster/audit/) recorded for the request.
+3. sets an `apiserver_requested_deprecated_apis` gauge metric to `1` in the `kube-apiserver`
```
Comment (celestehorgan) [inline content/en/docs/reference/using-api/deprecation-policy.md:305]: Non-blocking detail: capitalization 

```suggestion
1. Returns a `Warning` header (as defined in [RFC7234, Section 5.5](https://tools.ietf.org/html/rfc7234#section-5.5)) in the API response.
2. Adds a `"k8s.io/deprecated":"true"` annotation to the [audit event](/docs/tasks/debug-application-cluster/audit/) recorded for the request.
3. Sets an `apiserver_requested_deprecated_apis` gauge metric to `1` in the `kube-apiserver` 
```

**Comment by liggitt** [comment]
you can see what I pushed at https://github.com/liggitt/kubernetes-website/commits/warning-doc or wait for github to catch up here

**Comment by savitharaghunathan** [comment]
> I did push... looks like github is having issues...
> 
> @deads2k reviewed in [#21581 (comment)](https://github.com/kubernetes/website/pull/21581#issuecomment-654388559)

Can we get a thumbsup/lgtm from @deads2k ?
